Introduce EventBus

ChristianSchwarz opened the pull request:
**DO NOT MERGE YET!**

This is a prototype of an EventBus implementation that can be used by Mockito in the future to get rid of listener interfaces. The implementation follows the well known Guava EventBus.

Listener methods must have a method with a sole argument and must be marked with `@Subscribe`. It can be used like this

``` java
EventBus b = new EventBus();

b.register(new Object(){
    @Subscribe 
    public void log(MockitoEvent e){
        System.out.println(e);
    };
});

b.post(mockitoEvent);
```

There is much work to do, but please feel free to comment.

Todo's:
 - [ ] a healty API discussion
 - [ ] JavaDocs
 - [ ] provide transparent compability-mode for the currently known Mockito listener types
 - [ ] use this EventBus where ever possible
 - [ ] check Lambda compability
 - [ ] increase coverage
 - [ ] make EventBus Thread safe

szczepiq commented:
I am very excited about this idea!

Thanks for adding the code sample in the description - it really helps understanding the idea.

Without looking at the code, my immediate feedback is that annotation-driven APIs are harder to work with (discoverability, compile-time error reporting, implementation complexity).

TimvdLippe commented:
> The implementation follows the well known Guava EventBus.

Instead of reimplementing, could we take use/provide compatibility for their EventBus? Reduces maintenance cost for us :smile:

ChristianSchwarz commented:
> Instead of reimplementing, could we take use/provide compatibility for their EventBus? 
I inverstigated to "copy" the guava implementation. The guava solution is quite heavy (10 types) and has also dependencies to other guava classes that must be copied to or reimplemented too(MultiMap, Precondition, MoreExecutors, Objects, ...). Also the tests and the test tools have to be copied. An other point is the license, guava is released under Apache 2 and I have no clue if code can be transferred to an MIT project.

We can make the mockto event bus compatible with guava's @Subscribe annotation, but I don't know if it make much sense.

ChristianSchwarz commented:
Added backward compatible support to register for mock creation events:
 * in MockitoFramework I changed `addListener(MockitoListener)` to  `addListener(Object)`  it supports to register `MockCreationListener` as before or listeners that provide one or more subscriber methods. 
 * added a new event type `MockCreationReport` that must be used as sole parameter type in listener methods using the `@Subscribe` annotation

The _"old"_ way still works:

```java
Mockito.framework().addListener(new MockCreationListener(){
    public void onMockCreated(Object mock, MockCreationSettings settings){
        //jibbi it still works
    };
});
```

and the annotation based implementation (the listener method can have any visibility incl. private):

```java
Mockito.framework().addListener(new MockCreationListener(){
    @Subscribe void nameTheMethodAsYouLike(MockCreationReport report){
        //MockCreationReport includes the mock and the creation settings
    };
});
```

TimvdLippe commented on src/main/java/org/mockito/internal/progress/MockingProgress.java:
```diff
@@ -39,9 +42,30 @@
+     * Here is an example:
+     *  <pre>
+     *  Object listener =  new Object(){
+     *      public void logCreatedMocks(MockCreatedReport report){
```
The example is not annotated with `@Subscribe` :wink:

TimvdLippe commented on src/main/java/org/mockito/internal/progress/MockingProgress.java:
```diff
@@ -39,9 +42,30 @@
 
-    void removeListener(MockitoListener listener);
+    /**
+     * Unregisters the given listener. It will not be notified about mock creations anymore.
```
Mock creations is too specific, we plan to add more right?

ChristianSchwarz replied:
Yes, currently only MockCreationListener extends MockitoLister, other listener interfaces can be added in the same way.

TimvdLippe commented on src/main/java/org/mockito/internal/util/eventbus/EventBus.java:
```diff
@@ -0,0 +1,82 @@
+            }
+        }
+        
+        eventHandlers.removeAll(toBeRemoved);
```
Since `eventHandlers` is a `Set`, there can only be one?

ChristianSchwarz replied:
see my comment above

TimvdLippe commented on src/main/java/org/mockito/internal/util/eventbus/EventBus.java:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+    
+    public void unregister(Object listener) {
```
I think this whole method can be replaced with `this.eventHandlers.remove(listener)`

ChristianSchwarz replied:
Actually no, `eventHandlers` is a set of Subscriber instances, every Subscriber  represents a listener method of a listener instance. So every subscriber of the listener instance must be removed. The name `eventHandlers` is confusing.

If the mockito team decides to introduce an EventBus I will start more refactoring and polishing.

TimvdLippe commented on src/main/java/org/mockito/internal/util/eventbus/EventBus.java:
```diff
@@ -0,0 +1,82 @@
+
+        return true;
+    }
+
```
Unnecessary whitespace

TimvdLippe commented on src/main/java/org/mockito/internal/util/eventbus/Subscriber.java:
```diff
@@ -0,0 +1,78 @@
+        }
+    }
+
+    protected void invokeListenerMethod(Object event) {
```
This method feels very hacky. Especially since you are using reflection modify the accessibility. Would prefer a cleaner and nicer solution for this.

ChristianSchwarz replied:
Yes it is hacky, the error handling is also not final. In the current state the whole is a prototyp to see if it works at all. 

The idea is to allow methods and the declaring class to be non public. If the accessability is not changed we can't call public methods in private inner classes.

TimvdLippe commented:
To not lose type safety, maybe it is better to change `Object` to `MockitoListener` again and let `MockitoListener` implement default methods (requires Java 8 obv). Then whenever we want to add a new hook, we add a default empty method in `MockitoListener`. This is type-safe and lets users listen to multiple events at the same time.

ChristianSchwarz commented:
The question for me is, why do we need type safety here? What are the risks of the user. The addListener method can be implment more strict and throw is an instance is passed that implements no listener or has no subscriber methods. WDYT?

szczepiq commented:
Thank you for pushing the the idea forward!

At the moment, I don't want this implementation to be merged.

I'd like to step back and
a) identify what is the goal of the API change. What is the value of "getting rid of listener interfaces"?
b) evaluate alternative APIs, that don't use annotations.

Looking at the API in its current form, I don't want this to be merged because the API is:
 - harder to discover
 - fail in runtime instead of failing at compile time (type safety)
 - require writing a lot more code to handle
 - harder to work with and maintain in practice (in my experience of using them in Gradle)
 - annotations/reflection are typically slow (perhaps negligible but

Getting rid of listener interfaces (if that's the goal) has negative value. Mockito has and will have different families of listeners / subscriptions. Encoding this information in type hierarchy makes it easy to work with (cheap documentation, compiler checks).

szczepiq commented on src/main/java/org/mockito/MockitoFramework.java:
```diff
@@ -42,7 +42,7 @@
      * @since 2.1.0
      */
     @Incubating
-    MockitoFramework addListener(MockitoListener listener);
```
I reviewed the implementation and it seems like an overkill. What's the goal? See my comment to the entire PR.

ChristianSchwarz commented:
Thanks Szczepan and thanks Tim for your feedback!

About the general motivation: 
 * Annotation based event distribution allows the clients to name the listener method as they like. In traditional listener interface pattern clients are forced follow the method name of the listener interface. It is a good idea to name a method after its purpose. E.g:
```java
  @Subscribe
  private void sendVerificationFailureToBuildProcess(VerificationFailed event){
     ...
  }
``` 
* Annotation based event distribution allows to implement more than one event handling method per class. In an traditional approach you need to setup the chaining manually. 

```java
  @Subscribe
  private void sendToBuildProcess(VerificationFailed event){
     ...
  }

  @Subscribe
  private void logToConsole(VericationFailed event){
   ...
  }
``` 

 * The event bus allows to send event to private methods. The advantage here is that you don't have to write a private listener class and deleagte to a private method in order keep the visibility as low as possible. Example:

```java
  @Subscribe
  private void sendToBuildProcess(VerificationFailedEvent event){ ... }
``` 

vs:

```java
   //private or anonymous listeners are often used to implement  a listener iterface in favour of letting the top level classes implement it, cause this makes the public listener method a part of API of the top level class which is considered an anti pattern.
  private final VerificationFailedListener listener = new VerificationFailedListener(){
     public onVerificationFailed(VerificationFailedEvent event){
        sendToBuildProcess(event);
    }
  }

  private void sendToBuildProcess(VerificationFailedEvent event){ ...  }
``` 

* Events of the same family can be processed without the need to implement different listener methods and interfaces.

```java
  /* Assume  VerificationFailedEvent and VerificationSuccessEvent extend the base class VerificationEvent 
      both event types will be passed. 
  */
  @Subscribe
  private void sendToBuildProcess(VerificationEvent event){ ... }
```

ChristianSchwarz commented:
Benefits for mockio:
 * no need to to write sometimes redundant JavaDoc do for ListenerInterfaces and Events
 * same listener registration everywhere, e.g. currently some classes hold listeneres in a list where other hold the in a set
 * same error handling
 * Easy to discover API , we would have all events in one package and one annotation `@Subscribe` 
 * no listener interface to maintain 
 * no code duplication of listener iteration/notification/error handling

szczepiq commented:
@ChristianSchwarz, thanks a lot for taking the time and laying out the arguments!!!

I really like this idea and the pattern - but not for Mockito ;) It just seems like an overkill for our purposes. Also, you completely discount the cons of the proposed API (I guess following my strategy of discounting the benefits ;).

I don't feel too happy to curb your enthusiasm. I'd like to move on and stay with the current API for listeners. For now. Keep the ideas coming!!!

ChristianSchwarz commented:
Okay no problem! What should I do with this PR?
 * close
 * close and send again agains mockito 3
 * keep it open

bric3 commented:
I have the same opinion as @szczepiq on this. We already struggle to maintain existing code base. Having another one without substantial benefit (or benefit for a few) seems overkill. Nevertheless I'm not entirely closed to the idea.

Imho you should target master anyway. No need to open a new PR on this, as it's possible to change the target to master (maybe reserved to admins). Ping me if you can't.

TimvdLippe commented:
@ChristianSchwarz If you want to, I would be very interested if the default-method interface would work out. That would be type-safe and solves the duplication of listeners. Since you now target master, we can take advantage of Java 8 :smile:

szczepiq commented:
Let's close. We can always reopen :)

ChristianSchwarz commented:
@all Does it make sense to work on this idea for mockito 3?
@bric3 Currently  I can't change the target to master, maybe because this ticket is closed.

bric3 commented:
> @all Does it make sense to work on this idea for mockito 3?

That's not a priority, I think upgrading the codebase to Java 8 has more value.

@ChristianSchwarz I just changed it. Let's see if it's updated. 
@szczepiq Reopening temporarily

TimvdLippe commented:
> Does it make sense to work on this idea for mockito 3?

Agreed with @bric3 , we should reimplement core code for Java 8 first, then we can add new features on top of that.

Keep in mind a strong requirement is type-safety and up to this moment I am unconvinced you can develop an EventBus in a type-safe way
